src/mame/drivers/minitel_2_rpic.cpp

```diff
@@ -0,0 +1,475 @@
+				#endif
+			}
+
+			if( (port1 ^ data) & PORT_1_MDM_RTS )
+			{
+				#ifdef DEBUG_MINITEL
+				printf("PORT_1_MDM_RTS : %d \n", data & PORT_1_MDM_RTS );
+				#endif
+			}
+
+			if( (port1 ^ data) & PORT_1_KBLOAD )
+			{
+				#ifdef DEBUG_MINITEL
+				printf("PORT_1_KBLOAD : %d PC:0x%x\n", data & PORT_1_KBLOAD,m_maincpu->pc() );
+				#endif
+
+				if(data & PORT_1_KBLOAD)
+					keyboard_para_ser = 1;
+				else
+					keyboard_para_ser = 0;
+			}
+
+			if( (port1 ^ data) & PORT_1_SCL )
+			{
+				#ifdef DEBUG_MINITEL
+				printf("PORT_1_SCL : %d \n", data & PORT_1_SCL );
+				#endif
+			}
+
+			if( (port1 ^ data) & PORT_1_SDA )
+			{
+				#ifdef DEBUG_MINITEL
+				printf("PORT_1_SDA : %d \n", data & PORT_1_SDA );
+				#endif
+			}
+
+			port1=data;
+		break;
+
+		case MCS51_PORT_P2:
+			port2=data;
+		break;
+
+		case MCS51_PORT_P3:
+			port3=data;
+		break;
+	}
+}
+
+READ8_MEMBER(minitel_state::port_r)
+{
+	#ifdef DEBUG_MINITEL
+	printf("port_r: read PORT (offset=%02X) %x\n", offset,m_maincpu->pc());
+	#endif
+
+	switch(offset | 0x20000)
+	{
+		case MCS51_PORT_P0:
+			#ifdef DEBUG_MINITEL
+			printf("port_r: read %02X from PORT0\n", port0);
+			#endif
+			return port0;
+		case MCS51_PORT_P1:
+			#ifdef DEBUG_MINITEL
+			printf("port_r: read %02X from PORT1 - Keyboard -> %x\n", port1,((keyboard_x_row_reg>>7)&1));
+			#endif
+			return ( (port1&0xFE) | ((keyboard_x_row_reg>>7)&1) ) ;
+		case MCS51_PORT_P2:
+			#ifdef DEBUG_MINITEL
+			printf("port_r: read %02X from PORT2\n", port2);
+			#endif
+			return port2;
+		case MCS51_PORT_P3:
+			#ifdef DEBUG_MINITEL
+			printf("port_r: read %02X from PORT3\n", port3);
+			#endif
+			return port3;
+	}
+	return 0;
+}
+
+WRITE8_MEMBER(minitel_state::dev_crtl_reg_w)
+{
+	if( last_ctrl_reg != data)
+	{
+		#ifdef DEBUG_MINITEL
+		printf("minitel_state::hw_ctrl_reg : %x %x\n",offset, data);
+		#endif
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_DTMF )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_DTMF : %d \n", data & CTRL_REG_DTMF );
+			#endif
+		}
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_MCBC )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_MCBC : %d \n", data & CTRL_REG_MCBC );
+			#endif
+		}
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_OPTO )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_OPTO : %d \n", data & CTRL_REG_OPTO );
+			#endif
+		}
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_RELAY )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_RELAY : %d \n", data & CTRL_REG_RELAY );
+			#endif
+		}
+
+		if( (last_ctrl_reg ^ data) & CTRL_REG_CRTON )
+		{
+			#ifdef DEBUG_MINITEL
+			printf("CTRL_REG_CRTON : %d \n", data & CTRL_REG_CRTON );
+			#endif
+		}
+	}
+
+	last_ctrl_reg = data;
+}
+
+READ8_MEMBER(minitel_state::dev_keyb_ser_r)
+{
+	char kbdrow[8];
+
+	#ifdef DEBUG_MINITEL
+	printf("minitel_state::keyb read : %x\n",offset);
+	#endif
+
+	if ( keyboard_para_ser )
+	{
+		// load the 4014 with the keyboard row state
+		sprintf(kbdrow,"Y%.2X",offset>>8);
```

Comment:
Please use a `required_ioport_array` for this rather than a tagmap lookup on read.